include/swift/SIL/SILInstruction.h

```diff
@@ -591,6 +710,123 @@ inline SingleValueInstruction *SILNode::castToSingleValueInstruction() {
            inst->getKind() <= SILInstructionKind::Last_##ID;    \
   }
 
+/// A value base result of a multiple value instruction.
+class MultipleValueInstructionResult : public ValueBase {
+  /// The parent instruction that this MultipleValueInstructionResult is
+  /// contained within.
+  MultipleValueInstruction *Parent;
+
+  /// The ownership kind assigned to this result by its parent.
+  ///
+  /// We store this rather than delegating to our parent since there is not a
+  /// natural clear way to derive the ValueOwnershipKind for all potential
+  /// multiple value instructions. This contrasts with the index of our result
+  /// in the parent array.
+  ValueOwnershipKind Kind;
+
+public:
+  MultipleValueInstructionResult(MultipleValueInstruction *parent, SILType type,
+                                 ValueOwnershipKind kind)
+      : ValueBase(ValueKind::MultipleValueInstructionResult, type),
+        Parent(parent), Kind(kind) {}
+
+  MultipleValueInstruction *getParent() const { return Parent; }
+
+  unsigned getIndex() const;
```

Comment:
My recommendation would have been to store the index in the SubclassData (which should have a completely reasonable number of bits, even with the ValueOwnershipKind) and then store the parent-instruction pointer at the beginning of the array.  That might make it more difficult to use llvm::TrailingObjects, but I think it'll be fine.

Reply:
@rjmccall What do you mean by the beginning of the array? Which array? It can't be in the MultipleValueInstruction parent since we can't access that. Or am I missing something?

Reply:
I guess since I have the index, I can walk backwards assuming that I store the parent pointer. TBH though, couldn't I just not store the parent pointer at all and walk back using index and then walk back by an additional sizeof(MultipleValueInstruction)?

Reply:
I guess you would need to know the subclass as well. Hmmm... let me look at this a bit.

Reply:
Right.  My point is that you can always get back to the start of the array of MultiValueInstructionResults with just the current index, so you can store the parent pointer there without having to do any other instruction-specific dispatch.  But I guess you could also use instruction-specific dispatch if the start of the MultiValueInstructionResult array is always at a static offset (which would require it to always be the first trailing object).